flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,738 @@
+        weight_by (str): The method to weight the bins. Defaults to "min".
+        df_freq_pl (pl.DataFrame): A polars dataframe with the frequency of each bin.
+            Defaults to None.
+        percentiles (List[float]): The percentiles to for confidence intervals.
+            Defaults to [2.5, 97.5].
+        remove_any_null_turbine_bins (bool): When computing farm power, remove any bins where
+            and of the test turbines is null.  Defaults to False.
+        variance_only (bool): Only use the variance of turbine powers and not turbine-turbine
+            covariances. Defaults to False.
+        fill_cov_with_var (bool): Fill all missing covariance terms with the product of the
+            square root of the variances. Defaults to False.
+
+    Returns:
+        Dict[str, Dict[str, float]]: A dictionary containing the uplift results with standard error.
+    """
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+    #     print(df_sum)
+    #     print(uplift_results)
+
+    # Get the bin cols without df_name
+    bin_cols_without_df_name = [c for c in bin_cols_in if c != "df_name"]
+    bin_cols_with_df_name = bin_cols_in + ["df_name"]
+
+    # Add the wd and ws bins
+    df_ = _add_wd_ws_bins(
+        df_,
+        wd_cols=wd_cols,
+        ws_cols=ws_cols,
+        wd_step=wd_step,
+        wd_min=wd_min,
+        wd_max=wd_max,
+        ws_step=ws_step,
+        ws_min=ws_min,
+        ws_max=ws_max,
+    )
+
+    # Compute the covariance frame
+    df_cov = _compute_covariance(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    # If filling missing covariance terms, do it now
+    if fill_cov_with_var:
+        df_cov = _fill_cov_null(df_cov, test_cols=test_cols)
+
+    # If only using the variance, zero out the covariance terms
+    if variance_only:
+        df_cov = _zero_cov(df_cov, test_cols=test_cols)
+
+    # Apply Null values to covariance and sync across uplift pairs
+    df_cov = _null_and_sync_covariance(
+        df_cov=df_cov,
+        test_cols=test_cols,
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_cov)
+
+    # Bin and group the dataframe
+    df_bin = _bin_and_group_dataframe_expected_power(
+        df_=df_,
+        test_cols=test_cols,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # Synchronize the null values
+    df_bin = _synchronize_nulls(
+        df_bin=df_bin,
+        sync_cols=[f"{col}_mean" for col in test_cols],
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
+
+    # Remove rows where any of the test columns are null (if remove_any_null_turbine_bins is True)
+    if remove_any_null_turbine_bins:
+        df_bin = df_bin.filter(
+            pl.all_horizontal([pl.col(f"{c}_mean").is_not_null() for c in test_cols])
+        )
```

Comment:
Connected to the comment above, at this point, regardless of the value of "remove_any_null_turbine_bins," we might need to remove rows where all test_cols are Null. Although rare, I think it is possible that if count < 2 for all turbines in a bin, then after synchronizing Nulls between df_cov and df_bin, we could be left with rows that are all Null that should get filtered out from the analysis.

Reply:
Ok I think this is now in the line below